File: c/cwe121/002.c
// https://tgrez.github.io/posts/2022-06-19-buffer-overflow-in-rust.html
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void abracadabra() {
 printf("Abracadabra! Function called!\n");
 exit(0);
}

int main(int argc, char **argv) {
  struct {
    char buffer[64];
    volatile int (*point)();
  } hackvist;

  hackvist.point = NULL;
  strcpy(hackvist.buffer, argv[1]);

  printf("abracadabra function address: %p\n",abracadabra);
  printf("hackvist.point after strcpy: %p\n", hackvist.point);
  if (hackvist.point) {
    fflush(stdout);
    hackvist.point();
  } else {
    printf("Try Again\n");
  }

  exit(0);
}

File: rust/cwe121/002.rs
// https://tgrez.github.io/posts/2022-06-19-buffer-overflow-in-rust.html
use std::env;
use std::ffi::OsString;
use std::os::unix::ffi::OsStrExt;

fn abracadabra() {
    println!("Abracadabra! Function called!");
}

#[repr(C)]
struct Hackvist {
    buffer: [u8; 16],
    point: *const fn(),
}

fn main() {
    let mut args: Vec<OsString> = env::args_os().into_iter().collect();
    let first_arg: OsString = args.remove(1);
    let input_bytes: &[u8] = first_arg.as_bytes();
    let mut hackvist = Hackvist {
        buffer: [0; 16],
        point: 0 as *const fn() -> (),
    };

    unsafe {
        std::ptr::copy(
            input_bytes.as_ptr(),
            hackvist.buffer.as_mut_ptr(),
            input_bytes.len(),
        )
    }

    println!(
        "abracadabra function address: x{:0x}",
        abracadabra as *const () as usize
    );
    println!(
        "hackvist.point after strcpy: x{:0x}",
        hackvist.point as usize
    );
    println!(
        "hackvist.point after strcpy (in chars): {:?}",
        (hackvist.point as usize as u64)
            .to_le_bytes()
            .iter()
            .map(|b| *b as char)
            .collect::<String>()
    );

    if hackvist.point as usize == 0 {
        println!("Try again");
    } else {
        let code: fn() = unsafe { std::mem::transmute(hackvist.point) };
        code();
    }
}
